Uninitialized heap data returned in nocache mode when Reader::Read produces short read

Hi maintainers,

I found a correctness issue in the `Read()` path for `-o nocache` mode.

When `Reader::Read(offset, dst)` returns short (for example during sparse-hole reads), the code shrinks `dst` to the unread tail and then adds `dst.size()` to the return length. But the NUL padding line is currently commented out, so the unread tail bytes are never initialized before being counted as returned data.

Relevant code (`fuse-archive.cc`, around line 3398):

```cpp
ssize_t n = r->Read(offset, dst);
assert(n >= 0);
assert(n <= dst.size());
dst = dst.subspan(n);

// std::ranges::fill(dst, '\0');   // <-- disabled
assert(std::ranges::all_of(dst, [](char const c) { return c == '\0'; }));
n += dst.size();
return static_cast<int>(n);
```

In release builds (where `assert` is compiled out), this means the buffer tail goes back to the caller with whatever happened to be in heap memory.

The fix is straightforward: uncomment the `std::ranges::fill(dst, '\0')` line so that unread tail bytes are explicitly zeroed before inflating the return length.

I have a patch and a regression test ready. Happy to open a PR if that works for you.

Now, trying the test you provided in https://github.com/google/fuse-archive/pull/60 (but without the fix) on Ubuntu 24.04.4 LTS shows that this test does detect the issue:

```
$ make check
python3 test/test.py
...
INFO:root:Test sparse short read zero-fill in nocache mode
ERROR:root:Expected only NUL bytes for hole.bin, got non-zero bytes for size=4096 offset=0
ERROR:root:FAIL: There were 1 errors
make: *** [Makefile:38: check] Error 1

```

--- a/fuse-archive.cc
+++ b/fuse-archive.cc
@@ -3398,7 +3398,7 @@ int Read(const char*,
   // Pad the buffer with NUL bytes. This is a workaround for
   // https://github.com/libarchive/libarchive/issues/1194.
   // See https://github.com/google/fuse-archive/issues/40.
-  // std::ranges::fill(dst, '\0');
+  std::ranges::fill(dst, '\0');
   assert(std::ranges::all_of(dst, [](char const c) { return c == '\0'; }));
   n += dst.size();
 

--- a/test/test.py
+++ b/test/test.py
@@ -162,6 +162,7 @@ has_lzma = CanRun(['lzma', '--version'])
 has_lzop = CanRun(['lzop', '--version'])
 has_xz = CanRun(['xz', '--version'])
 has_zstd = CanRun(['zstd', '--version'])
+has_tar = CanRun(['tar', '--version'])
 
 sr = subprocess.run([mount_program, '--version'], capture_output=True, encoding='UTF-8')
 
@@ -1563,6 +1564,53 @@ def TestBigArchiveStreamed(options=[]):
             logging.debug(f'Unmounted {zip_path!r} from {mount_point!r}')
 
 
+# Regression test for sparse short reads in nocache mode.
+def TestNocacheSparseReadIsZeroFilled():
+    if is_fast or not has_tar: return
+    logging.info('Test sparse short read zero-fill in nocache mode')
+
+    with tempfile.TemporaryDirectory() as temp_dir:
+        payload_dir = os.path.join(temp_dir, 'payload')
+        os.mkdir(payload_dir)
+
+        with open(os.path.join(payload_dir, 'big.bin'), 'wb') as f:
+            f.write((b'SECRETSECRET|' * 90000)[:1024 * 1024])
+
+        with open(os.path.join(payload_dir, 'hole.bin'), 'wb') as f:
+            f.truncate(1024 * 1024)
+
+        archive_path = os.path.join(temp_dir, 'attack.tar')
+        subprocess.run(['tar', '--sparse', '-cf', archive_path, '-C', payload_dir, '.'], check=True)
+
+        mount_point = os.path.join(temp_dir, 'mnt')
+        os.mkdir(mount_point)
+        env = os.environ.copy()
+        env['MALLOC_PERTURB_'] = '170'
+        subprocess.run(
+            [mount_program, '-o', 'nocache,direct_io', '--', archive_path, mount_point],
+            check=True,
+            capture_output=True,
+            input='',
+            encoding='UTF-8',
+            env=env,
+        )
+        try:
+            fd = os.open(os.path.join(mount_point, 'hole.bin'), os.O_RDONLY)
+            try:
+                for offset in [0, 65536, 524288]:
+                    for size in [4096, 8192, 32768, 98304]:
+                        got = os.pread(fd, size, offset)
+                        if any(got):
+                            LogError(
+                                f'Expected only NUL bytes for hole.bin, got non-zero bytes '
+                                f'for size={size} offset={offset}'
+                            )
+            finally:
+                os.close(fd)
+        finally:
+            subprocess.run(['umount', '-l', mount_point], check=True)
+
+
 # Tests encrypted archive.
 def TestEncryptedArchive(options=[]):
     if not has_openssl and not has_nettle: return
@@ -1864,6 +1912,7 @@ TestDirectories()
 TestBigArchiveRandomOrder(['-o', 'direct_io'])
 TestBigArchiveRandomOrder(['-o', 'lazycache,direct_io'])
 TestBigArchiveStreamed(['-o', 'nocache,direct_io'])
+TestNocacheSparseReadIsZeroFilled()
 
 if error_count:
     LogError(f'FAIL: There were {error_count} errors')

